plugin/src/main/scala/chisel3/internal/plugin/SerializableModule.scala

```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.internal.plugin
+
+import scala.collection.mutable
+import scala.tools.nsc
+import scala.tools.nsc.{Global, Phase}
+import scala.tools.nsc.plugins.PluginComponent
+import scala.tools.nsc.symtab.Flags
+import scala.tools.nsc.transform.TypingTransformers
+
+// Create a object `SomeSerializableModuleGenerator` for `SomeSerializableModule extends SerializableModule[SomeSerializableModuleParameter]`:
+// object SomeSerializableModuleGenerator extends App {
+//   emitMlirbc(SerializableModuleGenerator(classOf[SomeSerializableModule], upickle.default.read[P](json)(implicitly[upickle.default.Read[P]])))
+// }
+class SerializableModuleGenerator(val global: Global, arguments: ChiselPluginArguments)
+    extends PluginComponent
+    with TypingTransformers
+    with ChiselOuterUtils {
+  import global._
+
+  val phaseName = "serializablemodulegeneratorphase"
+  val runsAfter: List[String] = List[String]("typer")
+  def newPhase(_prev: Phase): SerializableModuleGeneratorPhase = new SerializableModuleGeneratorPhase(_prev)
+
+  class SerializableModuleGeneratorPhase(prev: Phase) extends StdPhase(prev) {
+    override def name: String = phaseName
+    def apply(unit: CompilationUnit): Unit = {
+      if (ChiselPlugin.runComponent(global, arguments)(unit)) {
+        unit.body = new MyTypingTransformer(unit).transform(unit.body)
+      }
+    }
+  }
+
+  private class MyTypingTransformer(unit: CompilationUnit) extends TypingTransformer(unit) with ChiselInnerUtils {
+    override def transform(tree: Tree): Tree = tree match {
+      case packageDef @ PackageDef(pid, stats) =>
+        val newStats = stats.flatMap {
+          case classDef: ClassDef if isASerializableModule(classDef.symbol) && !classDef.mods.hasFlag(Flag.ABSTRACT) =>
```

Comment:
Rather than transforming at the package level, I suspect you can just do this at the `ClassDef` level and I think it would be best to insert the generated object into the companion object for the Class. We do that sort of thing in the `@instantiable` macro. Macros aren't exactly the same as plugins but there's a lot of similarity.

Reply:
> I think it would be best to insert the generated object into the companion object for the Class.

Here is of my original considerations:
- If adding a companion object, we may need to forbid user adding their own companion, this is possible, but I'm not sure if is necessary. or we need to implement a compile time logic to detect if there is already an companion, then adding logic to it, if not, creating a companion.
- This is the reason I choose to add a new object that user should never implement, e.g. `FooModuleGenerator`.

Reply:
> If adding a companion object, we may need to forbid user adding their own companion, this is possible, but I'm not sure if is necessary. or we need to implement a compile time logic to detect if there is already an companion, then adding logic to it, if not, creating a companion.

No need to forbid, you just create the object only if it doesn't exist, otherwise you just insert into the existing one. This is pretty standard practice (e.g. case classes generate methods in the companion object, creating it if necessary), it's a normal Scala-y thing to do so I would expect there to be an API.